Raghvendra eng patch 1

Raghvendra-eng opened the pull request.

Raghvendra-eng committed:
Added Transaction Class

Raghvendra-eng committed:
Update AllControlers.java

Added  getSummary controller

Raghvendra-eng committed:
Added interface for repolayer 

Handles transactions entity

Raghvendra-eng committed:
Service Layer for Transactions

Raghvendra-eng committed:
Update AccountServiceClass.java

Raghvendra-eng committed:
Changed for DAO Layer Repository name from "repo" to "repositories" and class name from "RepoLayer" to "UserRepository".

Raghvendra-eng committed:
Removed Auto Generated ToDo Comments and Commented Code

Raghvendra-eng committed:
Updates Repository name

Raghvendra-eng committed:
Updates Repository name and accordingly controllers class

Raghvendra-eng committed:
Commit for adding actuator and lombok dependencies

Raghvendra-eng committed:
Commit for adding actuator configuration

Raghvendra-eng committed:
Updates Repository name and accordingly controllers class

Raghvendra-eng committed:
lombok annotation added

Raghvendra-eng committed:
Added enum type to avoid hard coding of transactionType

Raghvendra-eng committed:
Changed long to Long. Added compareTo(). Added pathvariable validation. Added Response status to give proper message to user

Raghvendra-eng committed:
Added Exception Handling to Controllers

Raghvendra-eng committed:
Removed tomcat test method (/home api).

Raghvendra-eng committed:
User not found exception and redundancy issue fixes

Raghvendra-eng committed:
User not found exception and redundancy issue fixes and added Path variable validation

Raghvendra-eng committed:
Changes Added for Swagger Documentation

Abhinav-Cowkur commented on account/src/main/java/com/user/account/services/UserTransactionService.java:
```diff
@@ -21,7 +21,7 @@ public Transactions addTransaction(Transactions newTransaction) {
+	public ArrayList<Transactions> getSummary(Long accountNumber) {
 		return transactionRepository.findByAccountNumber(accountNumber);
 	}
 
```
Put the implementation classes in `services/impl` folder for better readability

Raghvendra-eng replied:
Resolved as recommended

Abhinav-Cowkur commented on account/src/main/java/com/user/account/controller/AllControlers.java:
```diff
@@ -50,55 +39,55 @@ public String home() {
+			if ( amount.compareTo(user.getAccountBalance()) <= 0 ) {
 				
-				user.incrementAccountBalance(-debitAmount) ;
+				user.incrementAccountBalance(-amount) ;
```
Have a `decrementAccountBalance` function to debit amount instead of using the same `incrementAccountBalance` function for debit and credit. Ideally, the increment and decrement functions should be in the service layer

Raghvendra-eng replied:
To avoid code redundancy decrementAccountBalance is not created.

nitsshukla-zz replied:
decrementAccountBalance is more readable and hence should be preferred.

You can save code redundancy by call `incrementAccountBalance (-amount)` inside `decrementAccountBalance` method

Raghvendra-eng replied:
Written separate incementAccountBalance and decrementAccountBalance method

Abhinav-Cowkur commented on account/src/main/java/com/user/account/controller/AllControlers.java:
```diff
@@ -21,73 +21,79 @@
+
 			User user = accountService.getUser(accountNumber);
-			
+			Long amount = Long.parseLong(transactionAmount);
```
Instead of parsing `transactionAmount` as Long datatype, obtain it as Long directly in the `debit` API, you can also have annotations to validate the transactionAmount (eg: negative value check)

Abhinav-Cowkur commented on account/src/main/java/com/user/account/controller/AllControlers.java:
```diff
@@ -21,73 +21,79 @@
+		
 			User user = accountService.getUser(accountNumber);
-				
+			Long amount = Long.parseLong(transactionAmount);
```
Same comment as https://github.com/Raghvendra-eng/account/pull/6#discussion_r607037594

nitsshukla-zz commented on account/src/main/java/com/user/account/services/TransactionService.java:
```diff
@@ -0,0 +1,12 @@
+	
+	public Transactions addTransaction(Transactions newTransaction);
+	
+	public ArrayList< Transactions> getSummary(Long accountNumber);
```
why `ArrrayList` and not `List`?

We should use abstract class everywhere it's possible

nitsshukla-zz commented on account/src/main/java/com/user/account/message/DefaultMessage.java:
```diff
@@ -0,0 +1,10 @@
+package com.user.account.message;
+
+public class DefaultMessage {
+	public final static String DEBIT = "Dabit";
```
`Dabit`?

nitsshukla-zz commented on account/src/main/java/com/user/account/controller/AllControlers.java:
```diff
@@ -1,65 +1,93 @@
+    public ResponseEntity<String> CreditAndDebitRequest(Long accountNumber, String transactionAmount, String transactionType){
+        try {
+            User user = accountService.getUser(accountNumber);
+            Long amount = Long.parseLong(transactionAmount);
```
This function is too complicated, please discuss with @Abhinav-Cowkur and simpify it.

High level pointers:

DEBIT/CREDIT can be `enum`
amount is already checked to be more than 0
I still don't see @Transactional in below code
```
transactionService.addTransaction(newTransaction);
accountService.updateUser(user);
```

Raghvendra-eng replied:
Added Nitin. It will come in today's commit

nitsshukla-zz commented on account/src/main/java/com/user/account/entity/Transactions.java:
```diff
@@ -0,0 +1,35 @@
+public class Transactions {
+	
+	@Id
+	private long transactionID;
```
why they are not `private final`?

Raghvendra-eng replied:
If I declare a variable as private final then i need to initialize it at the time of declaration or in the constructor. Since it is my id I can not initialize it at the time of declaration. And using Lombok for constructure, how to handle the error?

nitsshukla-zz replied:
Don't keep ID as final, make others final then.

Raghvendra-eng replied:
I need noArgConstructor for my implementation. If my variable will be a final variable, then i will not able to use noArgConstructor.

nitsshukla-zz replied:
Why'd you need `noArgConstructor `?

Raghvendra-eng replied:
noArgConstructor is needed for data ingestion into the table

nitsshukla-zz commented on account/src/main/java/com/user/account/entity/Transactions.java:
```diff
@@ -0,0 +1,35 @@
+@Table (name ="transactions" )
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter 
```
You can use @Data rather than so many annotations.

nitsshukla-zz commented on account/src/main/java/com/user/account/repositories/TransactionRepository.java:
```diff
@@ -0,0 +1,13 @@
+import com.user.account.entity.Transactions;
+
+public interface TransactionRepository extends JpaRepository<Transactions, Long>{
+	@Query
```
Why is this annotation being used here?

nitsshukla-zz commented on account/src/main/java/com/user/account/services/AccountServiceClass.java:
```diff
@@ -1,26 +1,27 @@
-		// TODO Auto-generated method stub
-		return repoLayerVari.getOne(accountNumber);
+	public User getUser(Long accountNumber) {
+		return userRepository.getOne(accountNumber);
```
If `accountNumber ` isn't present, what will happen?

Raghvendra-eng replied:
It will through EntityNotFoundException that is handled in catch block. User will get a message of invalid account number

nitsshukla-zz replied:
Through? Or Throw?

I don't think it's fine, EntityNotFound can happen in many places later on This approach is a bug prone in my opinion, you should throw a custom exception

Raghvendra-eng committed:
Changing transaction amount to Long

Raghvendra-eng committed:
@Data is added. removed @Getter and @Setter

Raghvendra-eng committed:
@Data is added. removed @Getter and @Setter for Transactions

Raghvendra-eng committed:
Typo error correction and added invalid account number message

Raghvendra-eng committed:
Replaced ArrayList with List

Raghvendra-eng committed:
code hygiene

Raghvendra-eng committed:
Replaced ArrayList with List

Raghvendra-eng committed:
Concurrency management using @Transactional

Raghvendra-eng committed:
code hygiene

Raghvendra-eng committed:
Replaced ArrayList with List

Raghvendra-eng committed:
Junit test cases

Raghvendra-eng committed:
Exception Handling and Code Hygiene Commit

Raghvendra-eng committed:
Unit testing for transaction Service

Raghvendra-eng committed:
Custom Exception Handling and code hygiene

nitsshukla-zz commented on account/src/main/java/com/user/account/controller/AllControlers.java:
```diff
@@ -1,65 +1,69 @@
+@Validated
 @RestController
+@RequestMapping("/")
 public class AllControlers {
```
spelling is not right for `Controller `?

nitsshukla-zz commented on account/src/main/java/com/user/account/controller/AllControlers.java:
```diff
@@ -1,65 +1,69 @@
+    // GetSummary for transactions in an account
+
+    @GetMapping("/getSummary/{accountNumber}")
+
```
don't have unnecessary empty lines please

nitsshukla-zz commented on account/src/test/java/com/user/account/services/AccountServiceClassTest.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Autowired
+    private AccountService accountService;
+
```
We should verify important mock calls, add a method 
```
@After
  public void tearDown() {
    verifyNoMoreInteractions(userRepository);
  }
```

nitsshukla-zz commented on account/src/test/java/com/user/account/services/AccountServiceClassTest.java:
```diff
@@ -0,0 +1,42 @@
+    @Test
+    public void updateUserTest() throws Exception{
+        User inputUser = new User(2, 1200, "Shyam");
+        //User outputUser = new User(2, 1100, "Shyam");
```
please no unused code as comment, remove it.

Hampers readability

nitsshukla-zz commented on account/src/test/java/com/user/account/services/UserTransactionServiceTest.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Autowired
+    private TransactionService transactionService;
+
```
verify as commented above.

nitsshukla-zz commented on account/src/main/java/com/user/account/controller/AllControlers.java:
```diff
@@ -1,65 +1,69 @@
+    @PostMapping("/debit/{accountNumber}")
+    public ResponseEntity<String> debit(@PathVariable @Min(1) Long accountNumber, @RequestParam @Min(1) Long amount) throws CustomExceptions {
+        User user = accountService.getUser(accountNumber);
+        if ( amount.compareTo(user.getAccountBalance()) <= 0){
```
such code should be in service layer not on controller layer, discuss it with peers.

Raghvendra-eng committed:
Added decrementAccountBalance, noMoreInteraction. Removed commented code, unusual line break.

Raghvendra-eng committed:
Removed unusual line break

Raghvendra-eng committed:
Removed unusual line break

Raghvendra-eng committed:
Added new API fund transfer from one user to another.

Abhinav-Cowkur commented on account/src/main/java/com/user/account/services/UserTransactionService.java:
```diff
@@ -1,37 +1,84 @@
+                    new Transactions(user.getAccountNumber(), amount, DefaultMessage.DEBIT);
+            addTransaction(newTransaction);
+            accountService.updateUser(user);
+            return new ResponseEntity<>(DefaultMessage.TRANSACTION_SUCCESS, HttpStatus.OK);
```
Just return the success message here, handle sending the `ResponseEntity` with status code in controller itself

Abhinav-Cowkur commented on account/src/main/java/com/user/account/services/UserTransactionService.java:
```diff
@@ -1,37 +1,84 @@
 
+    @Override
+    public ResponseEntity<String> debitService(User user, Long amount){
+        return debitImplementation(user, amount);
```
Why do you have a function just to call another function?

Abhinav-Cowkur commented on account/src/main/java/com/user/account/services/UserTransactionService.java:
```diff
@@ -1,37 +1,84 @@
+        ResponseEntity<String> returnResponse = new ResponseEntity<>(DefaultMessage.TRANSACTION_SUCCESS, HttpStatus.OK);
+        ResponseEntity<String> creditResponse;
+        if(!debitResponse.equals(returnResponse))
+            return debitResponse;
```
Instead of having a `returnResponse`, just check if `debitResponse` is Success and perform credit accordingly. The current implementation is a bit unreadable

Raghvendra-eng committed:
Moved service implementing class to new package under services class

Raghvendra-eng committed:
Moved HTTP status handler to controller layer.
Instead of using return Response compared with boolean in the debitService and creditService

Raghvendra-eng committed:
Added data validation and Exceptional handling for the new API

nitsshukla-zz commented on account/src/main/java/com/user/account/advice/AllControllerExceptionHandler.java:
```diff
@@ -0,0 +1,40 @@
+
+    @ExceptionHandler(value = ConstraintViolationException.class)
+    public ResponseEntity<String> validationErrorFound(){
+        return new ResponseEntity<>(DefaultMessage.VALIDATION_FAILED, HttpStatus.BAD_REQUEST);
```
`ConstraintViolationException` should not be 400, it can be 500 (`HttpStatus.INTERNAL_SERVER_ERROR`)

Raghvendra-eng replied:
Changed HttpStatus.BAD_REQUEST to HttpStatus.INTERNAL_SERVER_ERRO

nitsshukla-zz commented on account/src/main/java/com/user/account/advice/AllControllerExceptionHandler.java:
```diff
@@ -0,0 +1,40 @@
+
+    @ExceptionHandler(value = CustomExceptions.class)
+    public ResponseEntity<String> userNotFoundException(CustomExceptions c){
+        return new ResponseEntity<>(c.getMessage(), HttpStatus.BAD_REQUEST);
```
This also can be 501 or 500 or something, doesn't qualify for 400.

400 if request was badly provided

Raghvendra-eng replied:
Changed HttpStatus.BAD_REQUEST to HttpStatus.INTERNAL_SERVER_ERROR

nitsshukla-zz commented on account/src/main/java/com/user/account/advice/CustomExceptions.java:
```diff
@@ -0,0 +1,9 @@
+package com.user.account.advice;
+
+
+public class CustomExceptions extends Exception{
```
Very general name, try and give a specific name like `UserNotFoundException`

Raghvendra-eng replied:
Change as recommended

nitsshukla-zz commented on account/src/main/java/com/user/account/advice/CustomExceptions.java:
```diff
@@ -0,0 +1,9 @@
+package com.user.account.advice;
+
+
+public class CustomExceptions extends Exception{
```
Extend `RuntTimeException`, we can discuss if you should have Checked or unchecked

But here i think it should be unchecked.

Raghvendra-eng replied:
Extended custom exception with run time exception, Since haven't checked for the condition that provided account number exist or not.

nitsshukla-zz commented on account/src/main/java/com/user/account/controller/AllControllers.java:
```diff
@@ -0,0 +1,71 @@
+    private TransactionService transactionService;
+
+    //Send the response to the user based on transaction status
+    public ResponseEntity<String> transactionResponse(boolean transactionStatus){
```
Why is this method `public`?

Raghvendra-eng replied:
Change that method private, Since this method will not be called from anywhere outside of this class.

nitsshukla-zz commented on account/src/main/java/com/user/account/controller/AllControllers.java:
```diff
@@ -0,0 +1,71 @@
+            return new ResponseEntity<>(DefaultMessage.TRANSACTION_FAILED_LOW_BALANCE, HttpStatus.BAD_REQUEST);
+    }
+
+    // Updating accountBalance after debit
```
Don't put comments like this, prefer Javadoc

Raghvendra-eng replied:
Added JavaDoc documenattion

nitsshukla-zz commented on account/src/main/java/com/user/account/services/serviceImplementation/AccountServiceClass.java:
```diff
@@ -0,0 +1,27 @@
+    
+    @Override
+    public User updateUser(User user) {
+        return  userRepository.save(user);
```
[Minor] you've double space after `return`

Raghvendra-eng replied:
Remove a single space after return

nitsshukla-zz commented on account/build.gradle:
```diff
@@ -17,6 +23,18 @@ dependencies {
+	implementation 'org.springframework.boot:spring-boot-starter-actuator'
+	implementation 'org.springframework.boot:spring-boot-starter-data-rest'
+	compileOnly 'org.projectlombok:lombok'
+	developmentOnly 'org.springframework.boot:spring-boot-devtools'
```
why's this needed?

Raghvendra-eng replied:
The spring-boot-devtools module is included to provide additional development-time features. To include devtools support, add the module dependency to the build, Since it is not mandatory removed it for now.

nitsshukla-zz commented on account/build.gradle:
```diff
@@ -17,6 +23,18 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-web'
 	runtimeOnly 'mysql:mysql-connector-java'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+	compile group: 'org.mockito', name: 'mockito-core', version: '3.6.28' // JUnit dependencies
```
why's this needed as `compile ` and not `testCompile`?

Raghvendra-eng replied:
Changed the compile to testCompile

Raghvendra-eng committed:
CustomExceptions.java changed to UserNotFoundException.java
Change public transactionResponse method to private.
Changed HttpStatus.BAD_REQUEST to HttpStatus.INTERNAL_SERVER_ERROR

Raghvendra-eng committed:
Removed unused dependencies. (Spring boot devtools)

Raghvendra-eng committed:
Since Custom Exception is not checked changed Exception to RuntimeException

Raghvendra-eng committed:
Added Javadoc documentation comments
